fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -104,7 +108,22 @@
 
     private final LockFactory lockFactory;
     private final int blockSize;
-    private final Cache<String, FDBLuceneFileReference> fileReferenceCache;
+
+    /**
+     * A supplier used to load the {@link #fileReferenceCache}. The supplier allows us to memoize listing all
+     * the files from the database so that we don't have to do that more than once per {@link FDBDirectory}
```

Comment:
umm is it a correct assumption that the files list is always same within a transaction lifecycle? Like would it get updated when there are some writes in the transaction and you want to read your writes? and the segments merge can also lead to updated files if I understand correctly?

Reply:
You're right that the file list managed by a Lucene directory can change over the course of a transaction. To handle that, the cache is updated whenever a file is added, deleted, or renamed. For example, here: https://github.com/FoundationDB/fdb-record-layer/blob/70d252bbf8075c887f1614cb7376f19c57a4d969/fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java#L278

This change looks at all of the instances where the previous Guava cache (the `fileReferenceCache`) was updated, and now it makes sure that the full file cache has been loaded first, and then it uses that as the source of truth.

Reply:
In principle, it could be wasteful to load the full cache here rather than doing it lazily (which was the caching approach before), but the Lucene runtime was calling `listAll()` anyway (sometimes multiple times), and so in practice, I think it ends up being better to maintain a complete cache